loopd/swapclient_server_test.go

```diff
@@ -595,3 +599,170 @@ func TestHasBandwidth(t *testing.T) {
 		})
 	}
 }
+
+func TestListSwapsFilterAndPagination(t *testing.T) {
+	swapIn1 := loop.SwapInfo{
+		SwapStateData: loopdb.SwapStateData{
+			State: loopdb.StateInitiated,
+			Cost:  loopdb.SwapCost{},
+		},
+		LastUpdate:       time.Now(),
+		SwapHash:         lntypes.Hash{1},
+		SwapType:         swap.Type(swap.TypeIn),
+		HtlcAddressP2WSH: testnetAddr,
+		HtlcAddressP2TR:  testnetAddr,
+	}
+
+	swapOut1 := loop.SwapInfo{
+		SwapStateData: loopdb.SwapStateData{
+			State: loopdb.StateInitiated,
+			Cost:  loopdb.SwapCost{},
+		},
+		LastUpdate:       time.Now(),
+		SwapHash:         lntypes.Hash{2},
+		SwapType:         swap.Type(swap.TypeOut),
+		HtlcAddressP2WSH: testnetAddr,
+		HtlcAddressP2TR:  testnetAddr,
+	}
+
+	swapOut2 := loop.SwapInfo{
+		SwapStateData: loopdb.SwapStateData{
+			State: loopdb.StateInitiated,
+			Cost:  loopdb.SwapCost{},
+		},
+		LastUpdate:       time.Now(),
+		SwapHash:         lntypes.Hash{3},
+		SwapType:         swap.Type(swap.TypeOut),
+		HtlcAddressP2WSH: testnetAddr,
+		HtlcAddressP2TR:  testnetAddr,
+	}
+	// Create test cases with different filter/pagination parameters
+	tests := []struct {
+		name string
+		// Define the mock swaps that will be stored in the mock client
+		mockSwaps []loop.SwapInfo
+		// Request parameters
+		req *looprpc.ListSwapsRequest
+		// Expected response fields
+		expectedReturnedSwapCount  int
+		expectedFirstIdx           uint32
+		expectedLastIdx            uint32
+		expectedFilteredTotalCount uint32
+	}{
+		{
+			name:                       "fetch all swaps no pagination",
+			mockSwaps:                  []loop.SwapInfo{swapIn1},
+			req:                        &looprpc.ListSwapsRequest{},
+			expectedReturnedSwapCount:  1,
+			expectedFirstIdx:           0,
+			expectedLastIdx:            1,
+			expectedFilteredTotalCount: 1,
+		},
+		{
+			name:      "fetch swaps-ins no pagination",
+			mockSwaps: []loop.SwapInfo{swapIn1, swapOut1, swapOut2},
+			req: &looprpc.ListSwapsRequest{
+				ListSwapFilter: &looprpc.ListSwapsFilter{
+					SwapType: looprpc.ListSwapsFilter_LOOP_IN},
+			},
+			expectedReturnedSwapCount:  1,
+			expectedFirstIdx:           0,
+			expectedLastIdx:            1,
+			expectedFilteredTotalCount: 1,
+		},
+		{
+			name:                       "fetch swaps-outs no pagination",
+			mockSwaps:                  []loop.SwapInfo{swapIn1, swapOut1, swapOut2},
+			req:                        &looprpc.ListSwapsRequest{ListSwapFilter: &looprpc.ListSwapsFilter{SwapType: looprpc.ListSwapsFilter_LOOP_OUT}},
```

Comment:
Make sure that lines are not very long. We always try to format code such that we fit 80 columns (unless not possible). Please refer to our style guide: https://github.com/lightningnetwork/lnd/blob/master/docs/code_formatting_rules.md#additional-style-constraints